sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.{Condition, Lock, ReentrantLock}
+
+import org.apache.spark.sql.execution.QueryExecution
+import org.apache.spark.sql.util.QueryExecutionListener
+
+/**
+ * Not thread-safe.
+ * @param name
+ * @param sparkSession
+ */
+class Observation(name: String) {
```

Comment:
I'm thinking about what APIs we should provide here. We should at least have the `get` API to get the result and wait until it's available. We should also provide APIs with a timeout. Some ideas:
1. `def get(time: Long, unit: TimeUnit)`, which is similar to [LinkedBlockingQueue.poll](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/LinkedBlockingQueue.html#poll(long,%20java.util.concurrent.TimeUnit))
2. `waitCompleted` and users can call `get` later if it returns true.

It looks to me that `def get(time: Long, unit: TimeUnit)` should be good enough and people can do try-catch if they want to handle timeout differently. It's better to start with less public APIs as we can always add later.

Reply:
I find the `time` parameters more suitable with `waitCompleted`, and `get` should be the argument-less method.

Calling `get` without a successful `waitCompleted` should throw an `NoSuchElementException` or `IllegalStateException`.

What do you think about returning false on timeout and true on success? A timeout is not really an exceptional situation as it is something you can expect when you wait for something give a timeout.

Reply:
then shall we have a variant of `waitCompleted` that w/o a timeout?

Reply:
Yes, have done that now. The API is much thinner now.

Reply:
Well, a `waitCompleted` w/o a timeout is not really needed when we have `get` w/o a timeout. I have removed `reset` and even `close` as we use it only once.